### PR TITLE
Avoid unconditional `make_heap` for `priority_queue::push_range`

### LINKED ISSUE #2826: <queue>: priority_queue::push_range could conditionally push_heap
See discussion at https://github.com/microsoft/STL/pull/2806#discussion_r906814298.

When inserting a small number of elements `k` into a large `priority_queue`, `make_heap`'s complexity of `3N` can be greater than consecutive calls to `push_heap` (`k log N`). We need to determine what the actual relationship is given our implementation's constant factors, and further determine if we believe the potential perf improvement is worth the cost of additional codesize.

### PR DESCRIPTION
**Pull request by achabense**
Fixes #2826.

1. It's *dangerous* to unconditionally call `make_heap` - we will get great performance penalty if we do `push_heap` in small batches.
2. To remake the entire heap, for some types (like string), we can actually get better performance if we just do a `push_heap` loop. For those types, we can unconditionally do `push_heap` loop instead. However, for some trivial types (especially scalars) we will really get better performance if we call `make_heap`.
3. Based on 1&2, to be conservative, I'm choosing `new_size/2>old_size` as the boundary to prefer `make_heap` to `push_heap` loop. This will completely avoid the problem in 1.

About  `new_size/2>old_size`: `make_heap(size 2*o)` does `o` `_Pop_heap_hole_by_index`, and each `_Pop_heap_hole_by_index` calls a `_Push_heap_by_index`, so I think it's safe to assume doing a `o`-`push_heap` loop can be faster before/around this boundary. 

The following benchmark shows a generally better result, including those at the boundary (`Arg(vec_size / 2 + 1)`; 5001). However, I'm not sure going on in `u32/u64/float`'s `Arg(vec_size / 2 + 1)` case (in which case remaking the heap is actually faster). (This does not necessarily mean we have to set a smaller ratio boundary, as it looks size/type/sequence-sensitive. See the update part below; for more extensive benchmarks see the next comment.)

Benchmark result (`vec_size==10000`):
<details>
<summary>Previous</summary>

```
--------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations
--------------------------------------------------------------------------------------
BM_push_range<uint8_t, vec_u8>/1             132314020 ns    131250000 ns            5
BM_push_range<uint8_t, vec_u8>/100             1378146 ns      1380522 ns          498
BM_push_range<uint8_t, vec_u8>/10000             23991 ns        24065 ns        29867
BM_push_range<uint8_t, vec_u8>/5001              44356 ns        43493 ns        15448

BM_push_range<uint16_t, vec_u16>/1           129688300 ns    130208333 ns            6
BM_push_range<uint16_t, vec_u16>/100           1332057 ns      1339286 ns          560
BM_push_range<uint16_t, vec_u16>/10000           21871 ns        21973 ns        32000
BM_push_range<uint16_t, vec_u16>/5001            42268 ns        42375 ns        16593

BM_push_range<uint32_t, vec_u32>/1            49569970 ns     50000000 ns           10
BM_push_range<uint32_t, vec_u32>/100            534551 ns       531250 ns         1000
BM_push_range<uint32_t, vec_u32>/10000            8961 ns         8789 ns        74667
BM_push_range<uint32_t, vec_u32>/5001            23805 ns        24065 ns        29867

BM_push_range<uint64_t, vec_u64>/1            52815690 ns     53125000 ns           10
BM_push_range<uint64_t, vec_u64>/100            568480 ns       578125 ns         1000
BM_push_range<uint64_t, vec_u64>/10000           10845 ns        10882 ns        74667
BM_push_range<uint64_t, vec_u64>/5001            26797 ns        27274 ns        26353

BM_push_range<float, vec_float>/1             85530500 ns     85069444 ns            9
BM_push_range<float, vec_float>/100             902526 ns       899431 ns          747
BM_push_range<float, vec_float>/10000            16287 ns        16113 ns        40727
BM_push_range<float, vec_float>/5001             35487 ns        35296 ns        19478

BM_push_range<double, vec_double>/1           91232514 ns     91517857 ns            7
BM_push_range<double, vec_double>/100           931463 ns       927734 ns          640
BM_push_range<double, vec_double>/10000          18252 ns        17997 ns        37333
BM_push_range<double, vec_double>/5001           38361 ns        38365 ns        17920

BM_push_range<string_view, vec_str>/1        485819750 ns    484375000 ns            2
BM_push_range<string_view, vec_str>/100        5771718 ns      5781250 ns          100
BM_push_range<string_view, vec_str>/10000       128556 ns       128348 ns         5600
BM_push_range<string_view, vec_str>/5001        197701 ns       194972 ns         3446

BM_push_range<string, vec_str>/1             562048300 ns    562500000 ns            1
BM_push_range<string, vec_str>/100             6463913 ns      6556920 ns          112
BM_push_range<string, vec_str>/10000            160829 ns       160435 ns         4480
BM_push_range<string, vec_str>/5001             241398 ns       239955 ns         2800

BM_push_range<wstring_view, vec_wstr>/1      448094050 ns    445312500 ns            2
BM_push_range<wstring_view, vec_wstr>/100      6137002 ns      6250000 ns          100
BM_push_range<wstring_view, vec_wstr>/10000     125909 ns       125558 ns         5600
BM_push_range<wstring_view, vec_wstr>/5001      209610 ns       208575 ns         3446

BM_push_range<wstring, vec_wstr>/1           618522000 ns    625000000 ns            1
BM_push_range<wstring, vec_wstr>/100           8002004 ns      7986111 ns           90
BM_push_range<wstring, vec_wstr>/10000          452893 ns       449219 ns         1600
BM_push_range<wstring, vec_wstr>/5001           569804 ns       571987 ns         1120
```
</details>

<details>
<summary>Now</summary>

```
--------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations
--------------------------------------------------------------------------------------
BM_push_range<uint8_t, vec_u8>/1                129866 ns       128691 ns         4978
BM_push_range<uint8_t, vec_u8>/100               58166 ns        57812 ns        10000
BM_push_range<uint8_t, vec_u8>/10000             21505 ns        21484 ns        32000
BM_push_range<uint8_t, vec_u8>/5001              29584 ns        29820 ns        23579

BM_push_range<uint16_t, vec_u16>/1              134689 ns       136719 ns         5600
BM_push_range<uint16_t, vec_u16>/100             61244 ns        61384 ns        11200
BM_push_range<uint16_t, vec_u16>/10000           21685 ns        21484 ns        32000
BM_push_range<uint16_t, vec_u16>/5001            31650 ns        31390 ns        22400

BM_push_range<uint32_t, vec_u32>/1              133096 ns       131138 ns         5600
BM_push_range<uint32_t, vec_u32>/100             66086 ns        66964 ns        11200
BM_push_range<uint32_t, vec_u32>/10000           11868 ns        11440 ns        56000
BM_push_range<uint32_t, vec_u32>/5001            31735 ns        32087 ns        22400

BM_push_range<uint64_t, vec_u64>/1              137059 ns       134969 ns         4978
BM_push_range<uint64_t, vec_u64>/100             64009 ns        64174 ns        11200
BM_push_range<uint64_t, vec_u64>/10000           13127 ns        13393 ns        56000
BM_push_range<uint64_t, vec_u64>/5001            29801 ns        29820 ns        23579

BM_push_range<float, vec_float>/1               152839 ns       153460 ns         4480
BM_push_range<float, vec_float>/100              74416 ns        74986 ns         8960
BM_push_range<float, vec_float>/10000            15920 ns        16044 ns        44800
BM_push_range<float, vec_float>/5001             46340 ns        46038 ns        11200

BM_push_range<double, vec_double>/1             155410 ns       153460 ns         4480
BM_push_range<double, vec_double>/100            74945 ns        74986 ns         8960
BM_push_range<double, vec_double>/10000          21200 ns        21310 ns        34462
BM_push_range<double, vec_double>/5001           30883 ns        31145 ns        23579

BM_push_range<string_view, vec_str>/1           249675 ns       245536 ns         2800
BM_push_range<string_view, vec_str>/100         156717 ns       156948 ns         4480
BM_push_range<string_view, vec_str>/10000       131120 ns       131138 ns         5600
BM_push_range<string_view, vec_str>/5001        141407 ns       141246 ns         4978

BM_push_range<string, vec_str>/1                366666 ns       362739 ns         1723
BM_push_range<string, vec_str>/100              275281 ns       276215 ns         2489
BM_push_range<string, vec_str>/10000            162060 ns       160435 ns         4480
BM_push_range<string, vec_str>/5001             192471 ns       192540 ns         3733

BM_push_range<wstring_view, vec_wstr>/1         256913 ns       260911 ns         2635
BM_push_range<wstring_view, vec_wstr>/100       149590 ns       150663 ns         4978
BM_push_range<wstring_view, vec_wstr>/10000     125749 ns       125558 ns         5600
BM_push_range<wstring_view, vec_wstr>/5001      129565 ns       128691 ns         4978

BM_push_range<wstring, vec_wstr>/1              849736 ns       836680 ns          747
BM_push_range<wstring, vec_wstr>/100            716124 ns       714983 ns          896
BM_push_range<wstring, vec_wstr>/10000          488530 ns       497405 ns         1445
BM_push_range<wstring, vec_wstr>/5001           521849 ns       531250 ns         1000
```
</details>

<hr>

UPDATE: The new approach is actually faster around the boundary (`Arg(vec_size / 2 + 1)`; 4001,6001)  when `vec_size==8000 or 12000`:

When `vec_size==8000`:
<details>
<summary>Previous</summary>

```
-------------------------------------------------------------------------------------
Benchmark                                           Time             CPU   Iterations
-------------------------------------------------------------------------------------
BM_push_range<uint8_t, vec_u8>/1             86345400 ns     84821429 ns            7
BM_push_range<uint8_t, vec_u8>/100             896486 ns       899431 ns          747
BM_push_range<uint8_t, vec_u8>/8000             17466 ns        17648 ns        40727
BM_push_range<uint8_t, vec_u8>/4001             32089 ns        32227 ns        21333

BM_push_range<uint16_t, vec_u16>/1           82768829 ns     82589286 ns            7
BM_push_range<uint16_t, vec_u16>/100           880055 ns       878514 ns          747
BM_push_range<uint16_t, vec_u16>/8000           17378 ns        17648 ns        40727
BM_push_range<uint16_t, vec_u16>/4001           29446 ns        29157 ns        23579

BM_push_range<uint32_t, vec_u32>/1           66910836 ns     66964286 ns           14
BM_push_range<uint32_t, vec_u32>/100           713974 ns       711496 ns         1120
BM_push_range<uint32_t, vec_u32>/8000           19359 ns        19671 ns        37333
BM_push_range<uint32_t, vec_u32>/4001           33408 ns        32993 ns        20364

BM_push_range<uint64_t, vec_u64>/1           73805889 ns     74652778 ns            9
BM_push_range<uint64_t, vec_u64>/100           781004 ns       773929 ns          747
BM_push_range<uint64_t, vec_u64>/8000           22574 ns        22496 ns        29867
BM_push_range<uint64_t, vec_u64>/4001           38794 ns        39237 ns        17920

BM_push_range<float, vec_float>/1            79784767 ns     79861111 ns            9
BM_push_range<float, vec_float>/100            840182 ns       836680 ns          747
BM_push_range<float, vec_float>/8000            22378 ns        22321 ns        28000
BM_push_range<float, vec_float>/4001            39774 ns        39550 ns        16593

BM_push_range<double, vec_double>/1          85299433 ns     86805556 ns            9
BM_push_range<double, vec_double>/100          947088 ns       962182 ns          747
BM_push_range<double, vec_double>/8000          24678 ns        24554 ns        28000
BM_push_range<double, vec_double>/4001          44897 ns        44643 ns        11200

BM_push_range<string_view, vec_str>/1       535751600 ns    531250000 ns            1
BM_push_range<string_view, vec_str>/100       6278931 ns      6277902 ns          112
BM_push_range<string_view, vec_str>/8000       161811 ns       163923 ns         4480
BM_push_range<string_view, vec_str>/4001       267711 ns       263660 ns         2489

BM_push_range<string, vec_str>/1            587909800 ns    593750000 ns            1
BM_push_range<string, vec_str>/100            7078906 ns      6975446 ns          112
BM_push_range<string, vec_str>/8000            202050 ns       204041 ns         3446
BM_push_range<string, vec_str>/4001            318114 ns       320871 ns         2240

BM_push_range<wstring_view, vec_wstr>/1     579887500 ns    578125000 ns            1
BM_push_range<wstring_view, vec_wstr>/100     7370850 ns      7393973 ns          112
BM_push_range<wstring_view, vec_wstr>/8000     163061 ns       163923 ns         4480
BM_push_range<wstring_view, vec_wstr>/4001     278648 ns       276215 ns         2489

BM_push_range<wstring, vec_wstr>/1          786779400 ns    781250000 ns            1
BM_push_range<wstring, vec_wstr>/100         10046687 ns     10000000 ns           75
BM_push_range<wstring, vec_wstr>/8000          651075 ns       655692 ns         1120
BM_push_range<wstring, vec_wstr>/4001          468535 ns       468532 ns         1434
```
</details>

<details>
<summary>Now</summary>

```
-------------------------------------------------------------------------------------
Benchmark                                           Time             CPU   Iterations
-------------------------------------------------------------------------------------
BM_push_range<uint8_t, vec_u8>/1               108965 ns       109863 ns         6400
BM_push_range<uint8_t, vec_u8>/100              44074 ns        43316 ns        16593
BM_push_range<uint8_t, vec_u8>/8000             17109 ns        17264 ns        40727
BM_push_range<uint8_t, vec_u8>/4001             17287 ns        17264 ns        40727

BM_push_range<uint16_t, vec_u16>/1             112337 ns       112305 ns         6400
BM_push_range<uint16_t, vec_u16>/100            45185 ns        45409 ns        14452
BM_push_range<uint16_t, vec_u16>/8000           17384 ns        17648 ns        40727
BM_push_range<uint16_t, vec_u16>/4001           31069 ns        31250 ns        32000

BM_push_range<uint32_t, vec_u32>/1             157909 ns       160435 ns         4480
BM_push_range<uint32_t, vec_u32>/100            58340 ns        58594 ns        11200
BM_push_range<uint32_t, vec_u32>/8000           18087 ns        17997 ns        37333
BM_push_range<uint32_t, vec_u32>/4001           32847 ns        32087 ns        22400

BM_push_range<uint64_t, vec_u64>/1             160725 ns       160435 ns         4480
BM_push_range<uint64_t, vec_u64>/100            64445 ns        64174 ns        11200
BM_push_range<uint64_t, vec_u64>/8000           21812 ns        21973 ns        32000
BM_push_range<uint64_t, vec_u64>/4001           36642 ns        36901 ns        19478

BM_push_range<float, vec_float>/1              143998 ns       145777 ns         4073
BM_push_range<float, vec_float>/100             55574 ns        55804 ns        11200
BM_push_range<float, vec_float>/8000            11642 ns        11719 ns        64000
BM_push_range<float, vec_float>/4001            33503 ns        33692 ns        21333

BM_push_range<double, vec_double>/1            128889 ns       128348 ns         5600
BM_push_range<double, vec_double>/100           58495 ns        57199 ns        11200
BM_push_range<double, vec_double>/8000          12518 ns        12277 ns        56000
BM_push_range<double, vec_double>/4001          18326 ns        17997 ns        37333

BM_push_range<string_view, vec_str>/1          205341 ns       204041 ns         3446
BM_push_range<string_view, vec_str>/100        129671 ns       128691 ns         4978
BM_push_range<string_view, vec_str>/8000       103065 ns       104627 ns         7467
BM_push_range<string_view, vec_str>/4001       111280 ns       111607 ns         5600

BM_push_range<string, vec_str>/1               251210 ns       251116 ns         2800
BM_push_range<string, vec_str>/100             244371 ns       238550 ns         2358
BM_push_range<string, vec_str>/8000            135798 ns       134969 ns         4978
BM_push_range<string, vec_str>/4001            159636 ns       160435 ns         4480

BM_push_range<wstring_view, vec_wstr>/1        201277 ns       199507 ns         3446
BM_push_range<wstring_view, vec_wstr>/100      115028 ns       114397 ns         5600
BM_push_range<wstring_view, vec_wstr>/8000      96846 ns        96257 ns         7467
BM_push_range<wstring_view, vec_wstr>/4001      98729 ns        97656 ns         6400

BM_push_range<wstring, vec_wstr>/1             632818 ns       627790 ns         1120
BM_push_range<wstring, vec_wstr>/100           644742 ns       641741 ns         1120
BM_push_range<wstring, vec_wstr>/8000          381329 ns       383650 ns         1792
BM_push_range<wstring, vec_wstr>/4001          411618 ns       399013 ns         1723
```
</details>

When `vec_size==12000`:
<details>
<summary>Previous</summary>

```
--------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations
--------------------------------------------------------------------------------------
BM_push_range<uint8_t, vec_u8>/1             194486750 ns    195312500 ns            4
BM_push_range<uint8_t, vec_u8>/100             1976526 ns      1992754 ns          345
BM_push_range<uint8_t, vec_u8>/10000             69261 ns        66267 ns         8960
BM_push_range<uint8_t, vec_u8>/12000             35290 ns        35296 ns        19478
BM_push_range<uint8_t, vec_u8>/6001              59185 ns        58594 ns        11200

BM_push_range<uint16_t, vec_u16>/1           186077975 ns    187500000 ns            4
BM_push_range<uint16_t, vec_u16>/100           1906289 ns      1902174 ns          345
BM_push_range<uint16_t, vec_u16>/10000           65333 ns        65569 ns        11200
BM_push_range<uint16_t, vec_u16>/12000           34557 ns        34528 ns        20364
BM_push_range<uint16_t, vec_u16>/6001            61532 ns        61384 ns        11200

BM_push_range<uint32_t, vec_u32>/1           158355750 ns    156250000 ns            4
BM_push_range<uint32_t, vec_u32>/100           1635760 ns      1639230 ns          448
BM_push_range<uint32_t, vec_u32>/10000           68773 ns        69754 ns        11200
BM_push_range<uint32_t, vec_u32>/12000           38369 ns        38504 ns        18667
BM_push_range<uint32_t, vec_u32>/6001            61885 ns        61384 ns        11200

BM_push_range<uint64_t, vec_u64>/1           178653575 ns    179687500 ns            4
BM_push_range<uint64_t, vec_u64>/100           1854461 ns      1885054 ns          373
BM_push_range<uint64_t, vec_u64>/10000           79417 ns        79517 ns         7467
BM_push_range<uint64_t, vec_u64>/12000           43515 ns        43493 ns        15448
BM_push_range<uint64_t, vec_u64>/6001            68684 ns        68011 ns         8960

BM_push_range<float, vec_float>/1            189661800 ns    187500000 ns            4
BM_push_range<float, vec_float>/100            1948918 ns      1947464 ns          345
BM_push_range<float, vec_float>/10000            79855 ns        78474 ns         8960
BM_push_range<float, vec_float>/12000            43277 ns        43316 ns        16593
BM_push_range<float, vec_float>/6001             73937 ns        73242 ns         8960

BM_push_range<double, vec_double>/1          207170200 ns    203125000 ns            3
BM_push_range<double, vec_double>/100          2159209 ns      2148438 ns          320
BM_push_range<double, vec_double>/10000          89253 ns        90681 ns         8960
BM_push_range<double, vec_double>/12000          49647 ns        50000 ns        10000
BM_push_range<double, vec_double>/6001           77352 ns        76730 ns         8960

BM_push_range<string_view, vec_str>/1        745629400 ns    750000000 ns            1
BM_push_range<string_view, vec_str>/100        8897124 ns      8750000 ns           75
BM_push_range<string_view, vec_str>/10000       312743 ns       304813 ns         2358
BM_push_range<string_view, vec_str>/12000       167952 ns       170898 ns         4480
BM_push_range<string_view, vec_str>/6001        261305 ns       262277 ns         2800

BM_push_range<string, vec_str>/1             895937500 ns    890625000 ns            1
BM_push_range<string, vec_str>/100            10564659 ns     10602679 ns           56
BM_push_range<string, vec_str>/10000            384051 ns       383650 ns         1792
BM_push_range<string, vec_str>/12000            204698 ns       208575 ns         3446
BM_push_range<string, vec_str>/6001             317548 ns       313895 ns         2240

BM_push_range<wstring_view, vec_wstr>/1      711221700 ns    718750000 ns            1
BM_push_range<wstring_view, vec_wstr>/100      9292516 ns      9166667 ns           75
BM_push_range<wstring_view, vec_wstr>/10000     328228 ns       329641 ns         2133
BM_push_range<wstring_view, vec_wstr>/12000     160418 ns       160435 ns         4480
BM_push_range<wstring_view, vec_wstr>/6001      262532 ns       260911 ns         2635

BM_push_range<wstring, vec_wstr>/1           985596800 ns    968750000 ns            1
BM_push_range<wstring, vec_wstr>/100          12637839 ns     12555804 ns           56
BM_push_range<wstring, vec_wstr>/10000          831965 ns       837054 ns          896
BM_push_range<wstring, vec_wstr>/12000          587392 ns       585938 ns         1120
BM_push_range<wstring, vec_wstr>/6001           717141 ns       714983 ns          896
```
</details>

<details>
<summary>Now</summary>

```
--------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations
--------------------------------------------------------------------------------------
BM_push_range<uint8_t, vec_u8>/1                170014 ns       170898 ns         4480
BM_push_range<uint8_t, vec_u8>/100               79727 ns        80218 ns         8960
BM_push_range<uint8_t, vec_u8>/10000             35387 ns        35295 ns        20364
BM_push_range<uint8_t, vec_u8>/12000             30244 ns        30483 ns        23579
BM_push_range<uint8_t, vec_u8>/6001              51225 ns        50223 ns        11200

BM_push_range<uint16_t, vec_u16>/1              175976 ns       176467 ns         4073
BM_push_range<uint16_t, vec_u16>/100             79007 ns        77424 ns         7467
BM_push_range<uint16_t, vec_u16>/10000           25113 ns        25670 ns        28000
BM_push_range<uint16_t, vec_u16>/12000           30247 ns        29820 ns        23579
BM_push_range<uint16_t, vec_u16>/6001            41706 ns        41433 ns        16593

BM_push_range<uint32_t, vec_u32>/1              174957 ns       171611 ns         3733
BM_push_range<uint32_t, vec_u32>/100             84230 ns        85449 ns         8960
BM_push_range<uint32_t, vec_u32>/10000           19107 ns        19252 ns        37333
BM_push_range<uint32_t, vec_u32>/12000           21303 ns        20856 ns        34462
BM_push_range<uint32_t, vec_u32>/6001            43138 ns        43316 ns        16593

BM_push_range<uint64_t, vec_u64>/1              174656 ns       171611 ns         3733
BM_push_range<uint64_t, vec_u64>/100             82811 ns        81609 ns         7467
BM_push_range<uint64_t, vec_u64>/10000           18511 ns        18415 ns        40727
BM_push_range<uint64_t, vec_u64>/12000           23564 ns        23542 ns        29867
BM_push_range<uint64_t, vec_u64>/6001            38358 ns        37493 ns        17920

BM_push_range<float, vec_float>/1               192319 ns       192540 ns         3733
BM_push_range<float, vec_float>/100              95799 ns        96257 ns         7467
BM_push_range<float, vec_float>/10000            42719 ns        41992 ns        16000
BM_push_range<float, vec_float>/12000            32692 ns        32227 ns        21333
BM_push_range<float, vec_float>/6001             61729 ns        61384 ns        11200

BM_push_range<double, vec_double>/1             191856 ns       192540 ns         3733
BM_push_range<double, vec_double>/100            96711 ns        96257 ns         7467
BM_push_range<double, vec_double>/10000          31391 ns        31808 ns        23579
BM_push_range<double, vec_double>/12000          35254 ns        35296 ns        19478
BM_push_range<double, vec_double>/6001           54175 ns        54408 ns        11200

BM_push_range<string_view, vec_str>/1           311485 ns       313895 ns         2240
BM_push_range<string_view, vec_str>/100         200384 ns       199507 ns         3446
BM_push_range<string_view, vec_str>/10000       158855 ns       157286 ns         4073
BM_push_range<string_view, vec_str>/12000       150926 ns       149972 ns         4480
BM_push_range<string_view, vec_str>/6001        168579 ns       167411 ns         4480

BM_push_range<string, vec_str>/1                469091 ns       474330 ns         1120
BM_push_range<string, vec_str>/100              364140 ns       360947 ns         1948
BM_push_range<string, vec_str>/10000            244597 ns       239955 ns         2800
BM_push_range<string, vec_str>/12000            205429 ns       208575 ns         3446
BM_push_range<string, vec_str>/6001             246112 ns       245536 ns         2800

BM_push_range<wstring_view, vec_wstr>/1         320187 ns       320871 ns         2240
BM_push_range<wstring_view, vec_wstr>/100       207174 ns       208575 ns         3446
BM_push_range<wstring_view, vec_wstr>/10000     171600 ns       172631 ns         4073
BM_push_range<wstring_view, vec_wstr>/12000     160788 ns       163923 ns         4480
BM_push_range<wstring_view, vec_wstr>/6001      171280 ns       172631 ns         4073

BM_push_range<wstring, vec_wstr>/1             1046840 ns      1049805 ns          640
BM_push_range<wstring, vec_wstr>/100            921364 ns       927734 ns          640
BM_push_range<wstring, vec_wstr>/10000          625379 ns       627790 ns         1120
BM_push_range<wstring, vec_wstr>/12000          596649 ns       609375 ns         1000
BM_push_range<wstring, vec_wstr>/6001           622325 ns       627790 ns         1120
```
</details>

### CODE REVIEW

**Comment by achabense** [comment]
To show that `push_heap` loop can be faster than `make_heap` for remaking the heap:
(Pending)

**Comment by achabense** [comment]
To show that `push_heap` loop is a good strategy around `new_size/2~~old_size`:
(Pending)

File: benchmarks/src/priority_queue_push_range.cpp
Comment (achabense) [inline benchmarks/src/priority_queue_push_range.cpp:73]: `vec_size/2+1` will make the function firstly `make_heap` the first half of data, then `push_heap` the rest half of data.

File: stl/inc/xutility
Comment (achabense) [inline stl/inc/xutility:545]: Errr, this is not supported in C++14... can we use `noexcept(noexcept(_STD invoke(...)))` instead?

Comment (frederick-vs-ja) [reply stl/inc/xutility:545]: We can use `_Select_invoke_traits<_Fx&, _Args...>::_Is_nothrow_invocable::value` (`&&` can be omitted as it is added by default).

https://github.com/microsoft/STL/blob/d8ed02d2ec7ba8216848a33dec2b5fb385febf4e/stl/inc/type_traits#L1910

File: stl/inc/xutility
Comment (achabense) [inline stl/inc/xutility:546]: The two branches are both equivalent to `_STD invoke(_Fn, _STD forward<_Args>(_Vals)...);` can we remove the branch here?

**Comment by achabense** [comment]
> We can use _Select_invoke_traits

Hmm, sorry, I pushed too quickly

File: stl/inc/xutility
Comment (achabense) [reply stl/inc/xutility:546]: > We can use _Select_invoke_traits

Hmm, sorry, I pushed too quickly

File: stl/inc/xutility
Comment (CaseyCarter) [inline stl/inc/xutility:545]: No change required - I don't have data to back up the claim - but there's a lot of expensive SFINAE in `_Select_invoke_traits` that `_Ref_fn` doesn't need. We may be better off with:
```c++
template <class _Fx, bool = is_member_pointer_v<_Fx>>
struct _Ref_fn { // pass function object by value as a reference
    template <class... _Args>
    constexpr decltype(auto) operator()(_Args&&... _Vals) noexcept(
        noexcept(_STD invoke(_Fn, _STD forward<_Args>(_Vals)...))) {
        return _STD invoke(_Fn, _STD forward<_Args>(_Vals)...);
    }

    _Fx& _Fn;
};

template <class _Fx>
struct _Ref_fn<_Fx, false> {
    template <class... _Args>
    constexpr decltype(auto) operator()(_Args&&... _Vals) noexcept(noexcept(_Fn(_STD forward<_Args>(_Vals)...))) {
        return _Fn(_STD forward<_Args>(_Vals)...);
    }

    _Fx& _Fn;
};
```

File: benchmarks/src/priority_queue_push_range.cpp
Comment (StephanTLavavej) [inline benchmarks/src/priority_queue_push_range.cpp:14]: Style: We could use a digit separator here.

File: benchmarks/src/priority_queue_push_range.cpp
Comment (StephanTLavavej) [inline benchmarks/src/priority_queue_push_range.cpp:51]: Suggestion: `take_size` could be `const`.

File: benchmarks/src/priority_queue_push_range.cpp
Comment (StephanTLavavej) [reply benchmarks/src/priority_queue_push_range.cpp:40]: I realized that this might have been trying to stay below the SSO, but it doesn't for `wstring` which gets only 7 characters.

Comment (StephanTLavavej) [reply benchmarks/src/priority_queue_push_range.cpp:40]: I'll revert my change here since it could affect the benchmark behavior.

Comment (achabense) [reply benchmarks/src/priority_queue_push_range.cpp:40]: `make_heap/push_heap` attempts to move construct/assign variables. Casting to `u32` will result in `string` typically copying its characters and `wstring` just moving the pointer. Without this cast both will typically move their pointer. That said, this comparision might not makes much sense; and sorry for my lack of documentation.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I believe that this change is sufficiently safe, and the new benchmark provides justification for making the change, that we can merge this without a second maintainer approval.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for investigating and improving this performance with an added benchmark! :tada: :rocket: :stopwatch:

**Pull request merged**
